fenapack/field_split.py

```diff
@@ -15,159 +15,168 @@
 # You should have received a copy of the GNU Lesser General Public License
 # along with FENaPack.  If not, see <http://www.gnu.org/licenses/>.
 
-import dolfin
+"""This module provides subclasses of DOLFIN and petsc4py
+Krylov solvers implementing PCD fieldsplit preconditioned
+GMRES"""
+
+from __future__ import print_function
+
+from dolfin import Timer, PETScKrylovSolver
 from petsc4py import PETSc
 
 from fenapack._field_split_utils import dofmap_dofs_is
+from fenapack.field_split_backend import PCDInterface
+from fenapack.preconditioners import PCDPC_BRM1
+from fenapack.utils import get_default_factor_solver_package
+from fenapack.utils import allow_only_one_call
+
+
+class PCDKSP(PETSc.KSP):
+    """GMRES with right fieldsplit preconditioning using upper
+    Schur factorization and PCD Schur complement approximation.
+
+    This is a subclass of ``petsc4py.KSP``."""
 
-__all__ = ['FieldSplitSolver']
+    # NOTE: We are not able to overload PETSc.KSP methods in this
+    # class. That would only be possible with KSPPYTHON type but
+    # we do not need it.
 
+    def __init__(self, comm=None):
+        """Initialize PCDKSP on given MPI comm"""
 
-class FieldSplitSolver(dolfin.PETScKrylovSolver):
-    """This class derives from :py:class:`dolfin.PETScKrylovSolver` and
-    implements field split preconditioner for saddle point problems like
-    incompressible (Navier-)Stokes flow."""
-    def __init__(self, space, method, options_prefix=""):
-        """Create field split solver on a given space for a particular method.
+        super(PCDKSP, self).__init__()
 
-        *Arguments*
-            space (:py:class:`dolfin.FunctionSpace`)
-                Mixed function space determining the field split.
-            method (:py:class:`string`)
-                Type of a PETSc KSP object, see
-                help(:py:class:`petsc4py.PETSc.KSP.Type`).
+        self.create(comm)
+        self.setType(PETSc.KSP.Type.GMRES)
+        self.setPCSide(PETSc.PC.Side.RIGHT)
+        self.pc.setType(PETSc.PC.Type.FIELDSPLIT)
+        self.pc.setFieldSplitType(PETSc.PC.CompositeType.SCHUR)
+        self.pc.setFieldSplitSchurFactType(PETSc.PC.SchurFactType.UPPER)
+        self.pc.setFieldSplitSchurPreType(PETSc.PC.SchurPreType.USER)
+
+
+    @allow_only_one_call
+    def init_pcd(self, pcd_problem, pcd_pc_class=None):
+        """Initialize from ``PCDProblem`` instance. Needs to be called
+        after ``setOperators`` and ``setUp``. That's why two-phase
```

Comment:
I think it's about FIELDSPLIT PC `setUp`. But `KSP.setUp` will call this so it really does not matter which one it is. It is related to https://bitbucket.org/petsc/petsc/issues/160 and some demonstration of the behaviour can be seen in the unit test.